lib/copy.cpp

```diff
@@ -0,0 +1,164 @@
+                   c10::optional<at::Device> device = c10::nullopt,
+                   c10::optional<at::MemoryFormat> memory_format = c10::nullopt) {
+  TORCH_CHECK(x.layout() == at::Layout::Strided, "Only strided tensors are supported");
+  TORCH_CHECK(!x.is_quantized(), "Quantized tensors are not supported");
+
+  auto target_dtype = dtype.has_value() ? dtype.value() : x.scalar_type();
+  auto target_device = device.has_value() ? device.value() : x.device();
+  auto target_memory_format = memory_format.has_value() ? memory_format.value() : at::MemoryFormat::Preserve;
+
+  at::Tensor out =
+      at::empty_like(x, x.options().dtype(target_dtype).device(target_device), target_memory_format);
+
+  const int64_t numel = x.numel();
+  if (numel == 0) return out;
+
+  constexpr int BLOCK_SIZE = 1024;
+  const unsigned int grid_x = (numel + BLOCK_SIZE - 1) / BLOCK_SIZE;
+
+  c10::DeviceGuard guard(target_device);
+  c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
+  CUstream raw_stream = static_cast<CUstream>(stream.stream());
+
+  at::Tensor x_linear = (x.scalar_type() != target_dtype) ? x.to(target_dtype) : x;
+  if (x_linear.is_contiguous() && out.is_contiguous() && numel <= std::numeric_limits<int32_t>::max()) {
+    const TritonJITFunction& kernel_linear =
+        TritonJITFunction::get_instance((utils::get_triton_src_path() / "copy.py").string(),
+                                        "copy_kernel_linear");
+    kernel_linear(raw_stream, grid_x, 1, 1, 4, 0, x_linear, out, numel, BLOCK_SIZE);
+    return out;
+  }
+
+  std::vector<int64_t> task_shape(out.sizes().begin(), out.sizes().end());
+  int NDIMS = task_shape.size();
+
+  std::vector<int64_t> src_stride =
+      broadcasted_stride(std::vector<int64_t>(x_linear.sizes().begin(), x_linear.sizes().end()),
+                         std::vector<int64_t>(x_linear.strides().begin(), x_linear.strides().end()),
+                         task_shape);
+  std::vector<int64_t> dst_stride =
+      broadcasted_stride(std::vector<int64_t>(out.sizes().begin(), out.sizes().end()),
+                         std::vector<int64_t>(out.strides().begin(), out.strides().end()),
+                         task_shape);
+
+  const TritonJITFunction& kernel_nd =
+      TritonJITFunction::get_instance((utils::get_triton_src_path() / "copy.py").string(), "copy_kernel_nd");
+  kernel_nd(raw_stream,
+            grid_x,
+            1,
+            1,
+            4,
+            0,
+            x_linear,
+            out,
+            torch::tensor(task_shape, torch::TensorOptions().dtype(torch::kInt64).device(out.device())),
+            torch::tensor(src_stride, torch::TensorOptions().dtype(torch::kInt64).device(out.device())),
+            torch::tensor(dst_stride, torch::TensorOptions().dtype(torch::kInt64).device(out.device())),
+            numel,
+            NDIMS,
+            BLOCK_SIZE);
+
+  return out;
+}
+
+void copy_(const at::Tensor& dst, const at::Tensor& src) {
```

Comment:
the schema is  `Tensor&  copy_ (Tensor & self, const Tensor& src,  bool non_blocking)`